### PR TITLE
Initial webrtc and getUserMedia DOM support

### PR DESCRIPTION
**Pull request by Manishearth**
This is able to reach the point where connections are properly negotiated and ready to exchange streams.

<s>The `toJSON()` stuff doesn't work yet, so most example code will need to be tweaked to manually construct JSON first before sending SDP and ICE messages over websockets. I'll add support for this soon. (This may need webidl tweaks to support `[Default]` and `toJSON()`)</s>

For some reason I haven't yet figured out, connections are one-way, Servo is able to receive streams but the other end doesn't see the streams Servo sends. I don't think this is due to https://github.com/servo/media/issues/191, but that bug is making it harder to test.

This implementation simply drops streams that it receives, without connecting them up to any output elements, since servo-media-player doesn't yet have mediastream support.

Since servo can neither effectively send nor receive streams this implementation isn't useful yet, however it is getting large and I figured I'd get it reviewed and landed early as a base.

r? @jdm

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22780)

### CODE REVIEW

**Comment by Manishearth** [comment]
added toJSON support. Ideally `[Default]` would automatically work and give it to us for free for RTCSessionDescription, but writing it manually here isn't hard since we already support RTCSessionDescriptionInit.

**Comment by jdm** [comment]
```
./components/script/dom/webidls/RTCPeerConnection.webidl:91: no newline at EOF
```

File: components/script/dom/webidls/MediaDevices.webidl [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+[Exposed=Window,
+SecureContext, Pref="dom.webrtc.enabled"]
+interface MediaDevices : EventTarget {
+    //                attribute EventHandler ondevicechange;
+    // Promise<sequence<MediaDeviceInfo>> enumerateDevices();
+};
+
+partial interface Navigator {
+    // [SameObject, SecureContext]
+    readonly        attribute MediaDevices mediaDevices;
```
Comment (jdm) [inline components/script/dom/webidls/MediaDevices.webidl]: Pref-gate this.

File: components/script/dom/webidls/MediaDevices.webidl [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+partial interface MediaDevices {
+    // MediaTrackSupportedConstraints getSupportedConstraints();
+    Promise<MediaStream> getUserMedia(optional MediaStreamConstraints constraints);
```
Comment (jdm) [inline components/script/dom/webidls/MediaDevices.webidl]: Pref.

File: components/script/dom/webidls/RTCPeerConnectionIceEvent.webidl [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+// https://w3c.github.io/webrtc-pc/#rtcpeerconnectioniceevent
+
+[Constructor(DOMString type, optional RTCPeerConnectionIceEventInit eventInitDict),
+ Exposed=Window]
+interface RTCPeerConnectionIceEvent : Event {
```
Comment (jdm) [inline components/script/dom/webidls/RTCPeerConnectionIceEvent.webidl]: Pref.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
This all looks fine for the initial code dump, otherwise.

File: components/script/dom/webidls/MediaDevices.webidl (hunk 2 above)
Comment (Manishearth) [reply components/script/dom/webidls/MediaDevices.webidl]: no need, MediaDevices is pref-gated already

**Comment by jdm** [comment]
```
  = note: /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.1.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.1:function _ZN21servo_media_gstreamer13media_capture19create_input_stream17h7ba88d20259d4d8fE.llvm.5240714123248760377: error: undefined reference to 'gst_device_monitor_get_devices'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.1.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.1:function _ZN21servo_media_gstreamer13media_capture19create_input_stream17h7ba88d20259d4d8fE.llvm.5240714123248760377: error: undefined reference to 'gst_device_monitor_remove_filter'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.1.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.1:function _ZN21servo_media_gstreamer13media_capture19create_input_stream17h7ba88d20259d4d8fE.llvm.5240714123248760377: error: undefined reference to 'gst_device_get_caps'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.14.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.14:function _$LT$servo_media_gstreamer..webrtc..GStreamerWebRtcController$u20$as$u20$servo_media_webrtc..WebRtcControllerBackend$GT$::create_offer::h2c6582de0d5d514f: error: undefined reference to 'gst_promise_new_with_change_func'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.14.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.14:function _$LT$servo_media_gstreamer..webrtc..GStreamerWebRtcController$u20$as$u20$servo_media_webrtc..WebRtcControllerBackend$GT$::create_answer::h1b3755f9b5530baf: error: undefined reference to 'gst_promise_new_with_change_func'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.14.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.14:function servo_media_gstreamer::webrtc::GStreamerWebRtcController::set_description::h955e4325dd7cc0fc: error: undefined reference to 'gst_promise_new_with_change_func'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.7.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.7:function _$LT$O$u20$as$u20$gstreamer..auto..device..DeviceExt$GT$::create_element::h7f4184fb84d9a51d: error: undefined reference to 'gst_device_create_element'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.7.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.7:function _$LT$O$u20$as$u20$gstreamer..auto..device..DeviceExt$GT$::create_element::h7f4184fb84d9a51d: error: undefined reference to 'gst_device_create_element'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libservo_media_gstreamer-4a5ecd32a4dfe9b8.rlib(servo_media_gstreamer-4a5ecd32a4dfe9b8.servo_media_gstreamer.5fdvmycc-cgu.8.rcgu.o):servo_media_gstreamer.5fdvmycc-cgu.8:function _$LT$O$u20$as$u20$gstreamer..auto..device_monitor..DeviceMonitorExt$GT$::add_filter::h469fd1a1e6a65919: error: undefined reference to 'gst_device_monitor_add_filter'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libgstreamer-4c51a49ddff6d934.rlib(gstreamer-4c51a49ddff6d934.gstreamer.dten5zss-cgu.11.rcgu.o):gstreamer.dten5zss-cgu.11:function glib::translate::from_glib_full::h3e1f64e64102f54b: error: undefined reference to 'gst_device_monitor_get_type'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libgstreamer-4c51a49ddff6d934.rlib(gstreamer-4c51a49ddff6d934.gstreamer.dten5zss-cgu.11.rcgu.o):gstreamer.dten5zss-cgu.11:function glib::translate::from_glib_none::h4ab70408b3feb88c: error: undefined reference to 'gst_device_monitor_get_type'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libgstreamer-4c51a49ddff6d934.rlib(gstreamer-4c51a49ddff6d934.gstreamer.dten5zss-cgu.11.rcgu.o):gstreamer.dten5zss-cgu.11:function gstreamer::device_monitor::_$LT$impl$u20$gstreamer..auto..device_monitor..DeviceMonitor$GT$::new::h52821b98c8f3f7ce: error: undefined reference to 'gst_device_monitor_new'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libgstreamer-4c51a49ddff6d934.rlib(gstreamer-4c51a49ddff6d934.gstreamer.dten5zss-cgu.13.rcgu.o):gstreamer.dten5zss-cgu.13:function _$LT$gstreamer..auto..device..Device$u20$as$u20$glib..types..StaticType$GT$::static_type::h9d704d89e592a51a: error: undefined reference to 'gst_device_get_type'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libgstreamer-4c51a49ddff6d934.rlib(gstreamer-4c51a49ddff6d934.gstreamer.dten5zss-cgu.14.rcgu.o):gstreamer.dten5zss-cgu.14:function gstreamer::promise::Promise::get_reply::h5cb43c845af89946: error: undefined reference to 'gst_promise_get_reply'
          /home/servo/buildbot/slave/linux-rel-css/build/target/release/deps/libgstreamer-4c51a49ddff6d934.rlib(gstreamer-4c51a49ddff6d934.gstreamer.dten5zss-cgu.14.rcgu.o):gstreamer.dten5zss-cgu.14:function _$LT$gstreamer..promise..Promise$u20$as$u20$glib..types..StaticType$GT$::static_type::h38d774ccfd2b6a97: error: undefined reference to 'gst_promise_get_type'
          collect2: error: ld returned 1 exit status
```

**Comment by Manishearth** [comment]
looks like we need to update gstreamer on infra?

I'm on 1.14.1

cc @ferjm anyone know how to rebuild our tarballs for gst?

**Comment by jdm** [comment]
According to https://github.com/servo/saltfs/pull/937/files#diff-8f4416360af2ea20426022172cb18925R9 we should be using 1.14.3 already.

**Comment by ferjm** [comment]
> anyone know how to rebuild our tarballs for gst?

https://github.com/servo/servo/wiki/How-to-generate-GStreamer-binaries-for-Travis-(Ubuntu-Trusty) but I am not sure if that's the solution for [these errors](https://github.com/servo/servo/pull/22780#issuecomment-458692984). It seems that the undefined references are to functions that were introduced in 1.14.

**Comment by Manishearth** [comment]
@bors-servo retry

removed gst packages from the linux builders.

**Comment by Manishearth** [comment]
@bors-servo r=jdm

updated servo-media to include https://github.com/servo/media/pull/193

**Comment by Manishearth** [comment]
Looks like our android gst libraries don't have any gst_webrtc or sdp symbols.

But the pkg-config file says 1.14.3 so I'm not sure how this went wrong. @sdroege any idea what could be going on here?

The library being used is at https://servo-deps.s3.amazonaws.com/gstreamer/gstreamer-armeabi-v7a-1.14.3-20181105-103937.zip

**Comment by sdroege** [comment]
> But the pkg-config file says 1.14.3 so I'm not sure how this went wrong. @sdroege any idea what could be going on here?

I'm not sure if you're still using @ferjm's `libgstreamer_android_gen` thingy, but if you do the problem is here: https://github.com/servo/libgstreamer_android_gen/blob/536676168014b60d59ce8a0d1ef4003632d9c941/jni/Android.mk#L31

You have to add `gstreamer-sdp-1.0` and `gstreamer-webrtc-1.0` to that line to include all of the two libraries in the generated `libgstreamer_android.so`.

**Comment by Manishearth** [comment]
@sdroege i think we use https://github.com/servo/servo/wiki/How-to-generate-GStreamer-binaries-for-CI now ? which uses meson?

**Comment by Manishearth** [comment]
@ferjm could you have a look and regenerate binaries if we need to?

**Comment by sdroege** [comment]
> @sdroege i think we use https://github.com/servo/servo/wiki/How-to-generate-GStreamer-binaries-for-CI now ? which uses meson?

I believe that's only for the CI, not for the Android builds?

**Comment by ferjm** [comment]
> @ferjm could you have a look and regenerate binaries if we need to?

Sure.

> I believe that's only for the CI, not for the Android builds?

That's right.

**Comment by ferjm** [comment]
I built this branch locally with the new binaries and I was able to launch Servo (I wasn't able before updating the binaries as I was getting a `dlopen failed: cannot locate symbol "gst_webrtc_session_description_free"` error)

@bors-servo r=jdm

File: python/servo/build_commands.py [hunk 4]
```diff
@@ -441,7 +441,7 @@ def build(self, target=None, release=False, dev=False, jobs=None,
             # Build the name of the package containing all GStreamer dependencies
             # according to the build target.
             gst_lib = "gst-build-{}".format(self.config["android"]["lib"])
-            gst_lib_zip = "gstreamer-{}-1.14.3-20181105-103937.zip".format(self.config["android"]["lib"])
+            gst_lib_zip = "gstreamer-{}-1.14.3-20190131-153818.zip".format(self.config["android"]["lib"])
```
Comment (sdroege) [inline python/servo/build_commands.py]: Not sure if you did that, but you also should ensure that (in addition to the libraries themselves) the relevant plugins you're going to use are also included. For WebRTC that are going to be quite a few.

Comment (ferjm) [reply python/servo/build_commands.py]: Hmm, I only added gstreamer-sdp-1.0 and gstreamer-webrtc-1.0 to the list of plugins, as you suggested before. I'm not sure which plugins we need for WebRTC.

Comment (sdroege) [reply python/servo/build_commands.py]: You added those libraries to the list of libraries, nothing with plugins. The plugins are in a different variable.

I'm not sure what you all need, would have to check that too. At least webrtc, rtp, rtpmanager, vpx, opus and nice.

Comment (nirbheek) [reply python/servo/build_commands.py]: You need the following plugins: `opus vpx nice webrtc dtls srtp rtpmanager`. `sctp` too if/when you use data channels. I would recommend adding a runtime check that queries the gstreamer registry for the plugins that are available, something like this: https://github.com/centricular/gstwebrtc-demos/blob/master/sendrecv/gst-rust/src/main.rs#L113

Comment (ferjm) [reply python/servo/build_commands.py]: Ok, thanks!

We already have `opus` and `vpx`.

IIUC adding `GSTREAMER_PLUGINS_NET` to https://github.com/servo/libgstreamer_android_gen/blob/536676168014b60d59ce8a0d1ef4003632d9c941/jni/Android.mk#L30 should give us `webrtc dtls srtp` and `rtpmanager`. 

I don't see `nice` and `sctp` listed in `share/gst-android/ndk-build/plugins.mk` for 1.14.3.

Comment (nirbheek) [reply python/servo/build_commands.py]: `sctp` is not available in 1.14.x, but `nice` being missing is a bug. I'll fix it our build aggregator, but in the meantime you can add it to the `GSTREAMER_PLUGINS_NET` list in `plugins.mk` by hand for now. The plugin should already be packaged correctly.

**Comment by Manishearth** [comment]
Thanks everyone!

File: python/servo/build_commands.py (hunk 4 above)
Comment (nirbheek) [reply python/servo/build_commands.py]: The fix for `nice` has been [merged into 1.14](https://gitlab.freedesktop.org/gstreamer/cerbero/merge_requests/96) and will be available in 1.14.5, which should be out in the next few weeks.

Comment (ferjm) [reply python/servo/build_commands.py]: Great. Thank you! :)

**Pull request merged**